core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -540,39 +540,56 @@ class ReplicaManager(val config: KafkaConfig,
   }
 
   /*
-   * For each pair of partition and log directory specified in the map, record the pair in the memory so that the partition
-   * will be created in the specified log directory when broker receives LeaderAndIsrRequest for the partition later.
-   *
-   * This API is currently only useful if the replica has not been created yet. We will be able to move replicas
-   * that are already created to the user-specified log directory after KIP-113 is fully implemented
+   * For each pair of partition and log directory specified in the map, if the partition has already been created on
+   * this broker, move its log files to the specified log directory. Otherwise, record the pair in the memory so that
+   * the partition will be created in the specified log directory when broker receives LeaderAndIsrRequest for the partition later.
    *
    */
   def alterReplicaDir(partitionDirs: Map[TopicPartition, String]): Map[TopicPartition, Errors] = {
-    partitionDirs.map { case (topicPartition, destinationDir) =>
-      try {
-        if (!logManager.isLogDirOnline(destinationDir))
-          throw new KafkaStorageException(s"Log directory $destinationDir is offline")
-
-        // If the log for this partition has not been created yet:
-        // 1) Respond with ReplicaNotAvailableException for this partition in the AlterReplicaDirResponse
-        // 2) Record the destination log directory in the memory so that the partition will be created in this log directory
-        //    when broker receives LeaderAndIsrRequest for this partition later.
-        getReplica(topicPartition) match {
-          case Some(_) => // The support for moving replica between log directories on the same broker is not available yet.
-          case None =>
-            logManager.updatePreferredLogDir(topicPartition, destinationDir)
-            throw new ReplicaNotAvailableException(s"Replica $localBrokerId is not available for partition $topicPartition")
-        }
+    replicaStateChangeLock synchronized {
+      partitionDirs.map { case (topicPartition, destinationDir) =>
+        try {
+          if (!logManager.isLogDirOnline(destinationDir))
+            throw new KafkaStorageException(s"Log directory $destinationDir is offline")
+
+          // Stop current replica movement if the destinationDir is different from the existing destination log directory
+          getReplica(topicPartition, Request.FutureLocalReplicaId) match {
+            case Some(futureReplica) =>
+              if (futureReplica.log.get.dir.getParent != destinationDir) {
```

Comment:
Ah I forgot this. I just checked the KIP-113 wiki and it does say that broker will cancel existing movement if "any" is specified as the destination log directory.

I am wondering if we should to make the cancellation operation more explicit by saying, if "cancel" is specified as the destination log directory, the existing movement of this replica should be canceled. The issue with the existing design in the wiki is that user can not distinguish between "don't care" and "cancel existing movement". What do you think?

Reply:
Yes, explicitly modeling cancellation will be better. Perhaps we can change ALTER_REPLICA_DIR_REQUEST to sth like the following.

AlterReplicaDirRequest => [MovedPartitions] [CancelledPartitions]

MovedPartitions => logDir [topic [partitions]]
CancelledPartitions => [topic [partitions]]

With that, I am not sure we still need to support Any in logDir anymore.

Reply:
`Any` is only used in the json file that is provided to the `ReassignPartitionsCommand`. This is necessary in the case that user only wants to specify log directory for one out of three replicas for a given partition. In that case the `log_dirs` field in the json file will be something like `["any", "/path/to/logdir", "any"]`.

Also note that `ReassignPartitionsCommand` will filter out the `any` before it constructs AlterReplicaDirRequest. Thus `any` will not be specified as log directory in AlterReplicaDirRequest and broker does not need to understand this constant.

After more thinking, I think we probably don't need to provide a constant for user to cancel ongoing reassignment. There are two use-case for cancelling replica movement. One use-case is that user wants the replica to stay in the current log directory and he/she already knows the absolute path of the current log directory. In this case user can simply reassign partition again using the current log directory as the destination log directory. ReplicaManager will stop replica movement as appropriate.

The other use-case is that user wants to save broker IO by stop moving replica. In this case it is probably better to use quota to throttle replica movement. And user can also use either `kafka-log-dirs.sh` or `AdminClient.describeReplicaLogDir` to read the current log directory (and lag) of the partitions that are being moved, and reassign replica again with the current log directory as the destination log directory.

Thus it seems that we don't need to support explicit cancellation. Also, given that we are on a tight schedule to put this patch in 1.0 release, it may be better to try to reduce big changes to the patch unless it is necessary for the JBOD feature.

Reply:
Yes, we could make it simple for now that the only way to cancel is to specify the source log dir in AlterReplicaDirRequest.